### PR TITLE
Refactor handling of inherit_primary_header in AstroImage

### PR DESCRIPTION
**Pull request by ejeschke**
This PR refactors AstroImage, AstroTable and ginga.util.io_fits in the following ways:
- the primary header is saved as a metadata item called 'primary_header' (the header is already saved similarly as 'header'); this makes the handling of these two entities similar.
- the astropy/fitsio specific details for loading an HDU has been moved from AstroImage and AstroTable to ginga.util.io_fits. This encapsulates the details of the specific support package in io_fits
- added loading of FITS tables via the `fitsio` package in io_fits
- modified the TableView class to be able to view tables loaded with either astropy or fitsio

### CODE REVIEW

**Review by pllim — COMMENTED** [review]
I don't know how I feel about this. This is definitely a breaking change for `stginga`. What prompted this refactor?

**Comment by ejeschke** [comment]
@pllim, I certainly don't want to break things for `stginga`.  I'm just trying to clean up the "include primary header" implementation because it felt inconsistent with the way the regular headers are stored and handled. Could you let me know what breaks with this PR?  I'm am happy to modify it.

With this PR, the primary header is always saved, and is available from the image by `image.get('primary_header')`. The primary header is visible in the `Header` plugin by clicking the checkbox.

**Comment by pllim** [comment]
I'll have to double check. At the very least, there is a setting in `general.cfg`. I will also have to see if I call that affected API directly or not. I'll get back to you in a week or two. Thanks for your patience!

https://github.com/ejeschke/ginga/blob/04973bfe4e74d4c804abea7124d309b720b0a7a0/ginga/examples/configs/general.cfg#L51-L52

**Comment by ejeschke** [comment]
> Thanks for your patience!

No hurry on this one, @pllim!

File: ginga/AstroImage.py
Comment (pllim) [inline ginga/AstroImage.py:42]: I don't see `stginga` explicitly calling this in the constructor. Looks like `stginga` only sets it via `general.cfg` and you did not change the config name there, so that's good.

However, there is no guarantee that all your downstream packages/users also only set this via `general.cfg`. This is a public API change without a deprecation period. I am not sure if it is a good practice.

Comment (pllim) [reply ginga/AstroImage.py:42]: Recommendation: Instead of removing the keywords, deprecate them but preserve a backward-compatible functionality with deprecation warning, change log, and API documentation informing users and downstream devs. Wait for maybe two releases (your call on exactly how many) before removing them altogether.

Pro: No surprise for anyone.
Con: More work for you.

File: ginga/AstroImage.py
Comment (pllim) [inline ginga/AstroImage.py:66]: Same comment as https://github.com/ejeschke/ginga/pull/978/files#r776394464

File: ginga/AstroImage.py
Comment (pllim) [inline ginga/AstroImage.py:100]: Same comment as https://github.com/ejeschke/ginga/pull/978/files#r776394464 , except the warning here will be a bit different (renaming instead of removing).

`astropy` supports both cases: See https://github.com/astropy/astropy/blob/main/astropy/utils/decorators.py

File: ginga/AstroImage.py
Comment (pllim) [inline ginga/AstroImage.py:187]: Is this a behavior change, so would warrant mention in change log too but no need to undo the change.

File: ginga/AstroImage.py
Comment (pllim) [inline ginga/AstroImage.py:199]: Better be explicit here instead of relying on default setting, which might change in the future. One should never overwrite keyword in data extension from primary header.

```suggestion
                displayhdr.merge(hdr)
                displayhdr.merge(primary_hdr, override_keywords=False)
```

File: ginga/rv/Control.py
Comment (pllim) [inline ginga/rv/Control.py:100]: This behavior change also would need to be mentioned in change log.

File: ginga/table/AstroTable.py
Comment (pllim) [inline ginga/table/AstroTable.py:50]: Maybe use [specutils](https://github.com/astropy/specutils)?

File: ginga/util/io_fits.py
Comment (pllim) [inline ginga/util/io_fits.py:254]: Might also want to mention this in the change log. Addition might still break API if someone has been calling the keywords as arguments.

File: ginga/util/io_fits.py
Comment (pllim) [inline ginga/util/io_fits.py:344]: Same comment as https://github.com/ejeschke/ginga/pull/978/files#r776422501

File: ginga/util/io_fits.py
Comment (pllim) [inline ginga/util/io_fits.py:494]: Might want to deprecate this and have it return `copy_header(...)` instead of outright removal.

File: ginga/util/io_fits.py
Comment (pllim) [inline ginga/util/io_fits.py:565]: Same comment as https://github.com/ejeschke/ginga/pull/978/files#r776422501

File: ginga/util/io_fits.py
Comment (pllim) [inline ginga/util/io_fits.py:625]: Same comment as https://github.com/ejeschke/ginga/pull/978/files#r776422501

**Review by pllim — COMMENTED** [review]
I have some concerns about breaking change with no deprecation period (see comments above), but at a glance, I don't see this outright breaking `stginga`.

Therefore, it is okay to proceed and I will do an in-depth review once this is out of draft form. Would be nice to have deprecation but ultimately, it is your call.

If you decide to proceed with no deprecation period, you should bump the major version the release this patch is in.

**Comment by ejeschke** [comment]
Hi @pllim, can you have another look at this?  I have restored the non-private attributes and API of `AstroImage` and their former behavior.  I will submit a separate PR later for the deprecation of the items and we can discuss the issue further in that PR.

It is true that inherit_primary_header now defaults to `True` in `Control.py`, but that is a good thing, because in the previous version it did not matter at all whether it was True or False, the primary header was always saved.

**Comment by ejeschke** [comment]
Rebased

File: ginga/AstroImage.py
Comment (pllim) [inline ginga/AstroImage.py:206]: Do you intent to address the TODO as part of this PR?

File: ginga/AstroImage.py
Comment (pllim) [inline ginga/AstroImage.py:280]: Can this be simplified without actually trying to access the data that you never use?
```suggestion
        return 'primary_header' in self.metadata
```

**Review by pllim — COMMENTED** [review]
Still needs a change log.

I don't want to be a blocker of this getting merged. But I would want to have some time to test the release candidate with this change. Thanks for your patience!

**Comment by ejeschke** [comment]
> Still needs a change log.
> 
> I don't want to be a blocker of this getting merged. But I would want to have some time to test the release candidate with this change. Thanks for your patience!

No problem, @pllim.

File: ginga/AstroImage.py
Comment (ejeschke) [reply ginga/AstroImage.py:206]: No, that would change the exception type.  Will look at that for the follow-up PR.

File: ginga/AstroImage.py
Comment (ejeschke) [reply ginga/AstroImage.py:280]: I think I'd like to keep the type check on the value of the primary_header key. At least checking that it is not `None`.  We could check that it is a subclass of `Header`...

**Comment by ejeschke** [comment]
I will add some comments to the WhatsNew.

**Comment by ejeschke** [comment]
Rebased on master.

File: ginga/AstroImage.py
Comment (pllim) [reply ginga/AstroImage.py:280]: If you prefer to keep type check, I agree checking for `Header` might be more explicit. 😸

File: doc/WhatsNew.rst
Comment (pllim) [inline doc/WhatsNew.rst:32]: Maybe this will fix `doc/WhatsNew.rst:36: ERROR: Unexpected indentation`?
```suggestion
  responsible for loading those files:
```

Comment (ejeschke) [reply doc/WhatsNew.rst]: Looked at an example further down in the file.  Looks like I needed the colon and perhaps a blank line for RST to recognize the sublist.

File: doc/WhatsNew.rst
Comment (pllim) [inline doc/WhatsNew.rst:29]: Fixing this is out of scope, perhaps, but it needs fixing too.
```suggestion

  - choice of sampling by grid now; useful for mosaics and collages
  - for previous parameter of usecrop=True, use sample=crop
```

Comment (ejeschke) [reply doc/WhatsNew.rst:30]: Good catch!

**Review by pllim — APPROVED** [review]
Can't keep this open forever... Feel free to merge. Thanks! 😸

**Comment by ejeschke** [comment]
> Can't keep this open forever... Feel free to merge. Thanks! smile_cat

Thanks for the great review, @pllim. If you notice anything amiss we can fix it before the next release.

**Pull request merged**
